### DISCUSSION TITLE
[11 Settembre 2023] Modifica CPU

### DISCUSSION THREAD

**Discussion by FeddyLix17** [sapienzastudentsnetwork/architettura-degli-elaboratori#75] [Modifiche architettura MIPS (M-Z)]
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/d068dd7d-0f30-463f-a2c4-fbc28efc244e)
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/cce4ffa3-ee4c-4923-a0d1-df108ef1c2ff)

**Comment by FeddyLix17** [comment]
## 1) Modifiche all'architettura

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/5f63c12c-fc86-4aab-91c4-134f751393c6)

## 2) Word che esprime l'istruzione *vjc $9, $10, 0x0FFA*

- l'Opcode dell'istruzione vjc (fornito in esadecimale dalla traccia) è uguale a $\Large 0x3A_{16}  =  0011\ 1010_2$ , codificato in 6 bit 111010

- il campo $rs, rappresentate il primo registro sorgente, è $9, codificato in 5 bit come 01001(registro $t1)

- il campo $rt, rappresentante il secondo registro sorgente, è $10, codificato in 5 bit come 01010 (registro numero $t2)

- il campo immediate, rappresentante l'indirizzo del primo elemento del vettore, è $\Large 0x0FFA_{16}$, codificato in 16 bit come 0000_1111_1111_1010

![photo_2023-10-26_12-50-28](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/3561dded-bec4-4075-91be-ae37ddfb9d29)

## 3) Durata totale ciclo di clock

Considerando

- la fase di *Instruction Fetch*, nella quale si accederà alla memoria, con tempo *75 ns*

- la fase di *Instruction Decode*, nella quale si accederà ai registri con tempo *25 ns* e in parallelo inizierà l'estensione del segno di instr[15-0]

- la fase di *Execute*, nella quale verrano parallelamente

    - calcolata la differenza dei valori contenuti nei registri $rs e $rt
    
    - finito di estendere il segno dell'immediate ed in parte iniziato lo shift logico di 2 di quest'ultimo

    - calcolata la somma lunghezza vettore -1 (cosi da avere l'indice dell'ultimo elemento)
    con tempo *100 ns* 

- una volta finito lo shift logico, ci sarà una quarta fase nella quale verrano calcolati i valori
    - indirizzo primo elemento vettore + (lunghezza vettore -1)
    - indirizzo primo elemento vettore + $rs (ovvero $indice)
 
    con tempo *100 ns*

la durata totale di un ciclo di clock tale da permettere l'esecuzione della nuova istruzione dovrà essere di almeno

$$
    \Large 75 + 25 + 100 + 75 + 100 = 375 ns
$$

<div align="center">

![photo_2023-10-26_16-00-45](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/b295deef-a3e5-48a9-9d3a-291f16f0727a)

</div>

## 4) Control Unit malfunzionante

Sapendo che

- tutti i segnali di tipo don't care siano pari a 0

- e che la Control Unit difettosa sovrascriva il segnale RegDst come segue:

$$ \Large RegDst = vjc$$

si analizza il comportamento del seguente frammento di codice

```assembly
addi $s1, $0, 5
addi $s0, $s0, 20
sub $s0, $s0, $s1
```
assumendo che $s0 e $s1 siano inizializzati a 0

- $\Large \text{addi \\$s1, \\$0, 5}$

al termine dell'istruzione il registro $s1 dovrebbe contenere il valore $0 (registro $at) +5,
- $\Large \text{addi \\$s0, \\$s0, 20}$

al termine dell'istruzione il registro $s0 viene correttamente incrementato di 20

- $\Large \text{sub \\$s0, \\$s0, \\$s1}$

al termine dell'istruzione, il registro $s0 dovrebbe contenere la differenza tra se stesso e il valore del registro $r1 (20 - (
$at +5), ma essendo (difettosamente) impostato a 0 il segnale RegDst il risultato verrà (erroneamente) salvato nel registro $s1

Al termine dell'esecuzione del frammento di codice, i valori di $s0 e $s1 saranno rispettivamente 20 e  20 - ($at + 5).

**Reply by thegeek-sys** [reply]
Ho qualche dubbio sul funzionamento della tua CPU modificata.
1. Non dovrebbe essere shiftato di 2 l'indice del vettore invece dell'indirizzo del vettore?
2. Il valore non dovrebbe essere letto dalla memoria prima di effettuare il salto?

Io ho risolto così (spero sia comprensibile)
<img width="1055" alt="Screenshot 2024-06-05 alle 10 56 24" src="https://github.com/sapienzastudentsnetwork/architettura-degli-elaboratori/assets/55345633/ef6c7189-a88e-4e24-901b-127fa862a4df">

**Reply by FeddyLix17** [reply]
In quale punto non vengono rispettate queste 2 condizioni?

**Reply by thegeek-sys** [reply]
nella tua CPU non leggi nulla dalla memoria (tanto che MemRead è impostato a zero). Per la traccia dell'esercizio il dato letto corrisponde ad un indirizzo a cui bisogna saltare nel caso in cui vengano rispettate le condizioni.

Inoltre per come è impostato l'esercizio, il vettore è composto da word. Il che vuol dire che per leggere l'indice-simo elemento del vettore, prima di fare la somma, bisogna moltiplicare l'indice per 4 (shiftarlo) e solo dopo eseguire la somma con l'indirizzo base del vettore (tu invece shifti l'indirizzo del vettore)

<img width="1072" alt="immagine" src="https://github.com/sapienzastudentsnetwork/architettura-degli-elaboratori/assets/55345633/36e447ac-3306-4dc2-bce7-010de0a503e6">

**Reply by FeddyLix17** [reply]
prendi quello che sto per dire con le pinze in quanto sono mesi che non tocco più questa materia 🙏🏻 

per il punto 2 della tua risposta, l'indirizzo a cui saltare (ma in generale le istruzioni di tipo jump) di solito non richiedono un accesso in memoria (per come è il tuo disegno è come se gli stessi dicendo di saltare al valore contenuto all'indirizzo a cui vogliamo saltare, invece che all'indirizzo stesso), in altre parole dopo aver calcolato l'indirizzo lo passiamo subito al selettore (aggiuntivo) del program counter.

per il punto 1 (e la parte finale) della tua risposta invece, è corretto affermare che essendo un vettore di word gli indici vadano moltiplicati per 4 (shiftati di 2), e trovo sempre corretto affermare che ad essere shiftato nel mio disegno sarebbe dovuto essere $indice e non vettore

**Comment by thegeek-sys** [comment]
@FeddyLix17 il registro `$0` se non sbaglio non è `$at` (che è il registro numero 1) bensì è il registro che contiene di default il vaore 0

**Reply by FeddyLix17** [reply]
Si, hai ragione.

Comunque il comportamento generale dovrebbe rimanere invariato
